R/provider-snowflake.R

```diff
@@ -0,0 +1,184 @@
+#'   variable.
+#' - Key-pair authentication credentials defined via the `SNOWFLAKE_USER` and
+#'   `SNOWFLAKE_PRIVATE_KEY` (which can be a PEM-encoded private key or a path
+#'   to one) environment variables.
+#' - Posit Workbench-managed Snowflake credentials for the corresponding
+#'   `account`.
+#'
+#' @inheritParams chat_openai
+#' @inheritParams chat_cortex
+#' @inherit chat_openai return
+#' @examplesIf has_credentials("cortex")
+#' chat <- chat_snowflake()
+#' chat$chat("Tell me a joke in the form of a SQL query.")
+#' @export
+chat_snowflake <- function(system_prompt = NULL,
+                           turns = NULL,
+                           account = snowflake_account(),
+                           credentials = NULL,
+                           model = NULL,
+                           api_args = list(),
+                           echo = c("none", "text", "all")) {
+  turns <- normalize_turns(turns, system_prompt)
+  check_string(account, allow_empty = FALSE)
+  model <- set_default(model, "llama3.1-70b")
+  echo <- check_echo(echo)
+
+  if (is_list(credentials)) {
+    static_credentials <- force(credentials)
+    credentials <- function(account) static_credentials
+  }
+  check_function(credentials, allow_null = TRUE)
+
+  provider <- ProviderSnowflake(
+    base_url = snowflake_url(account),
+    account = account,
+    credentials = credentials,
+    model = model,
+    extra_args = api_args,
+    # We need an empty api_key for S7 validation.
+    api_key = ""
+  )
+
+  Chat$new(provider = provider, turns = turns, echo = echo)
+}
+
+ProviderSnowflake <- new_class(
+  "ProviderSnowflake",
+  parent = ProviderOpenAI,
+  properties = list(
+    account = prop_string(),
+    credentials = class_function | NULL
+  )
+)
+
+# See: https://docs.snowflake.com/en/user-guide/snowflake-cortex/cortex-llm-rest-api#api-reference
+method(chat_request, ProviderSnowflake) <- function(provider,
+                                                    stream = TRUE,
+                                                    turns = list(),
+                                                    tools = list(),
+                                                    type = NULL,
+                                                    extra_args = list()) {
+  if (length(tools) != 0) {
+    cli::cli_abort(
+      "Tool calling is not supported for {.code chat_snowflake()}."
+    )
+  }
+  if (!is.null(type) != 0) {
+    cli::cli_abort(
+      "Structured data extraction is not supported for {.code chat_snowflake()}."
+    )
+  }
+  if (!stream) {
+    cli::cli_abort(
+      "Non-streaming responses are not supported for {.code chat_snowflake()}."
+    )
+  }
+
+  req <- request(provider@base_url)
+  req <- req_url_path_append(req, "/api/v2/cortex/inference:complete")
+  creds <- cortex_credentials(provider@account, provider@credentials)
+  req <- req_headers(req, !!!creds, .redact = "Authorization")
```

Comment:
I wondered about always redacting `Authorization` in httr2. What do you think?

Reply:
That kind of makes sense to me, yeah.